File: app/src/main/java/at/bitfire/icsdroid/AppAccount.java
/*
 * Copyright (c) 2013 – 2016 Ricki Hirner (bitfire web engineering).
 *
 * This program is free software: you can redistribute it and/or modify it under the terms of the GNU
 * General Public License as published by the Free Software Foundation, either version 3 of the
 * License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful, but WITHOUT ANY WARRANTY;
 * without even the implied warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR
 * PURPOSE.  See the GNU General Public License for more details.
 *
 */

package at.bitfire.icsdroid;

import android.accounts.Account;
import android.accounts.AccountManager;
import android.content.ContentResolver;
import android.content.Context;
import android.content.PeriodicSync;
import android.os.Bundle;
import android.provider.CalendarContract;
import android.util.Log;

public class AppAccount {
    public static final long SYNC_INTERVAL_MANUALLY = -1;

    public static final Account account = new Account("ICSdroid", "at.bitfire.icsdroid");


    public static void makeAvailable(Context context) {
        AccountManager am = AccountManager.get(context);
        if (am.getAccountsByType(account.type).length == 0) {
            Log.i(Constants.TAG, "Account not found, creating");
            am.addAccountExplicitly(AppAccount.account, null, null);
            ContentResolver.setIsSyncable(account, CalendarContract.AUTHORITY, 1);
            ContentResolver.setSyncAutomatically(account, CalendarContract.AUTHORITY, true);
        }
    }

    public static boolean isSyncActive() {
        return ContentResolver.isSyncActive(AppAccount.account, CalendarContract.AUTHORITY);
    }

    public static long getSyncInterval(Context context) {
        makeAvailable(context);

        long syncInterval = SYNC_INTERVAL_MANUALLY;
        if (ContentResolver.getSyncAutomatically(account, CalendarContract.AUTHORITY))
            for (PeriodicSync sync : ContentResolver.getPeriodicSyncs(account, CalendarContract.AUTHORITY))
                syncInterval = sync.period;
        return syncInterval;
    }

    public static void setSyncInterval(long syncInterval) {
        if (syncInterval == -1) {
            Log.i(Constants.TAG, "Disabling automatic synchronization");
            ContentResolver.setSyncAutomatically(account, CalendarContract.AUTHORITY, false);
        } else {
            Log.i(Constants.TAG, "Setting automatic synchronization with interval of " + syncInterval + " seconds");
            ContentResolver.setSyncAutomatically(account, CalendarContract.AUTHORITY, true);
            ContentResolver.addPeriodicSync(account, CalendarContract.AUTHORITY, new Bundle(), syncInterval);
        }
    }
}

File: app/src/main/java/at/bitfire/icsdroid/AppAccount.kt
/***************************************************************************************************
 * Copyright © All Contributors. See LICENSE and AUTHORS in the root directory for details.
 **************************************************************************************************/

package at.bitfire.icsdroid

import android.accounts.Account
import android.accounts.AccountManager
import android.content.ContentResolver
import android.content.Context
import android.content.SharedPreferences.OnSharedPreferenceChangeListener
import android.provider.CalendarContract
import android.util.Log
import at.bitfire.icsdroid.worker.PeriodicSyncWorker
import kotlinx.coroutines.channels.Channel
import kotlinx.coroutines.channels.awaitClose
import kotlinx.coroutines.flow.buffer
import kotlinx.coroutines.flow.callbackFlow

object AppAccount {

    const val DEFAULT_SYNC_INTERVAL = 24*3600L   // 1 day
    const val SYNC_INTERVAL_MANUALLY = -1L

    private const val PREF_ACCOUNT = "account"
    private const val KEY_SYNC_INTERVAL = "syncInterval"

    private var account: Account? = null


    @Synchronized
    fun get(context: Context): Account {
        // use cached account, if available
        account?.let {
            return it
        }

        // get account
        val accountType = context.getString(R.string.account_type)

        val am = AccountManager.get(context)
        val existingAccount = am.getAccountsByType(accountType).firstOrNull()
        if (existingAccount != null) {
            account = existingAccount
            return existingAccount
        }

        Log.i(Constants.TAG, "Account not found, creating")
        val newAccount = Account(context.getString(R.string.account_name), accountType)
        if (am.addAccountExplicitly(newAccount, null, null)) {
            account = newAccount
            ContentResolver.setIsSyncable(newAccount, CalendarContract.AUTHORITY, 1)
            setSyncInterval(context, DEFAULT_SYNC_INTERVAL)
            return newAccount
        }

        throw IllegalStateException("Couldn't create app account")
    }


    fun getSyncIntervalFlow(context: Context) = callbackFlow {
        val preferences = getPreferences(context)
        val listener = OnSharedPreferenceChangeListener { _, key ->
            if (key == KEY_SYNC_INTERVAL)
                trySend(preferences.getLong(KEY_SYNC_INTERVAL, DEFAULT_SYNC_INTERVAL))
        }
        preferences.registerOnSharedPreferenceChangeListener(listener)
        if (preferences.contains(KEY_SYNC_INTERVAL))
            send(preferences.getLong(KEY_SYNC_INTERVAL, DEFAULT_SYNC_INTERVAL))
        awaitClose {
            preferences.unregisterOnSharedPreferenceChangeListener(listener)
        }
    }.buffer(Channel.UNLIMITED)

    fun setSyncInterval(context: Context, syncInterval: Long) {
        // don't use the sync framework anymore (legacy)
        ContentResolver.setSyncAutomatically(get(context), CalendarContract.AUTHORITY, false)

        // remember sync interval so that it can be checked/restored later
        getPreferences(context).edit()
            .putLong(KEY_SYNC_INTERVAL, syncInterval)
            .apply()

        // set up periodic worker
        PeriodicSyncWorker.setInterval(
            context,
            if (syncInterval == SYNC_INTERVAL_MANUALLY)
                null
            else
                syncInterval
        )
    }


    // helpers

    private fun getPreferences(context: Context) =
        context.getSharedPreferences(PREF_ACCOUNT, 0)

}
